dbms_test.go

```diff
@@ -0,0 +1,163 @@
+package sqllexer
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type output struct {
+	Expected          string             `json:"expected"`
+	ObfuscatorConfig  *obfuscatorConfig  `json:"obfuscator_config,omitempty"`
+	NormalizerConfig  *normalizerConfig  `json:"normalizer_config,omitempty"`
+	StatementMetadata *StatementMetadata `json:"statement_metadata,omitempty"`
+}
+
+type outputList []*output
+
+func getSubdirectories(directory string) ([]string, error) {
+	var subdirs []string
+	entries, err := os.ReadDir(directory)
+	if err != nil {
+		return nil, err
+	}
+	for _, entry := range entries {
+		if entry.IsDir() {
+			subdirs = append(subdirs, entry.Name())
+		}
+	}
+	return subdirs, nil
+}
+
+// TestQueriesPerDBMS tests a preset of queries and expected output per DBMS
+// Test folder structure:
+// -- testdata
+//
+//	-- dbms_type
+//	  -- query_type
+//	    -- query_name.sql
+//	    -- query_name.expected
+func TestQueriesPerDBMS(t *testing.T) {
+	dbmsTypes := []DBMSType{
+		DBMSPostgres,
+	}
+
+	for _, dbms := range dbmsTypes {
+		// Get all subdirectories of the testdata folder
+		baseDir := filepath.Join("testdata", string(dbms))
+		// Get all subdirectories of the testdata folder
+		queryTypes, err := getSubdirectories(baseDir)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, qt := range queryTypes {
+			dirPath := filepath.Join(baseDir, qt)
+			files, err := os.ReadDir(dirPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			for _, file := range files {
+				// the testdata folder contains only .sql and .expected files
+				// so we can safely ignore the other files
+				if strings.HasSuffix(file.Name(), ".sql") {
+					// Remove the .sql extension to get the test name
+					testName := strings.TrimSuffix(file.Name(), ".sql")
+					t.Run(testName, func(t *testing.T) {
+						queryPath := filepath.Join(dirPath, file.Name())
+						expectedPath := filepath.Join(dirPath, testName+".expected")
+
+						input, err := os.ReadFile(queryPath)
+						if err != nil {
+							t.Fatal(err)
+						}
+
+						expectedJson, err := os.ReadFile(expectedPath)
+						if err != nil {
+							t.Fatal(err)
+						}
+
+						var expectedOutputs outputList
+
+						if err := json.Unmarshal(expectedJson, &expectedOutputs); err != nil {
```

Comment:
I worry a bit that we lose type safety on those definitions of expected outputs by having them defined purely as json. I'm not sure how big of a deal that is but having the expected outputs as go structs instead does have the benefit of being refactoring-friendly (renaming or removing a field can be taken care of by the IDE instead of requiring search and replace).

Reply:
The reason I chose `json` over `go` struct is because
1. I would like the test files under `testdata` folder to be pure text, and the expected output file to have a non-standard `.expected` file extension so it's clear to developers. 
2. The `expected` output are unmarshalled to an [output](https://github.com/DataDog/go-sqllexer/pull/25/files#diff-0aa4d72fd772473fdbf70e31a0d1dcc1a9ff9c33311207e97cfc373a5c2e50dbR13-R18) type so technically we still have the static typing. 
3. In case we want to move the huge amount of test files into a different repo, text based testdata will actually make the migration easier.

Reply:
> The expected [output](https://github.com/DataDog/go-sqllexer/pull/25/files#diff-0aa4d72fd772473fdbf70e31a0d1dcc1a9ff9c33311207e97cfc373a5c2e50dbR13-R18) are unmarshalled to an output type so technically we still have the static typing.

You do have type validation but you're still losing the static type safety during refactoring. Maybe not a big deal in this case if we stick to the format (or even if we don't, it might be fine to deal with the search and replace in the files). 

I don't feel strongly about it.

Reply:
If I do change the test expected output to a `.go` file, am I still able to load the respective expected output dynamically at runtime? Or I need to look for some other techniques like code generation and `init` func to load the expected outputs into a map?

Reply:
I think what I'm describing is essentially what we're doing in [tests like these](https://github.com/DataDog/dd-go/blob/prod/database-monitoring/apps/dbm-metadata-processor/decoder/decoder_test.go#L21-L27). In those, we start with some file content and then have table driven tests that validate that the output is the typed struct value that we expect. 

It's not dynamic but I don't know if that's a requirement. The workflow wouldn't be that different either:
1. Add an input sql file
2. Define the expected result for your test case
    1. In your version, that means an expected json file
    2. In the alternative, this means a new test case value that gives the name of the input file and defined the expected result struct value.

In both instances, we've got a table driven test but one has the test cases inferred and generated dynamically from files on disk and the other just has them explicitly listed in a slice of test cases. I'll also note that we do have a variation in `dbm-metrics-processor` where we have a set of input/expected files like you've got here. I think everything is a tradeoff and you can even see hints in the comment on that [metrics-processor test](https://github.com/DataDog/dd-go/blob/prod/database-monitoring/apps/dbm-metrics-processor/processor/metrics_test.go#L546-L614). 

So, as you can see, I've written both versions. Each one for different reasons and scope. I mostly just wanted to have us pick the right tradeoff for this case. I think the smell I can see in this case is that the expected json file maps to a struct that we define in tests rather than just some output json that another component or system consumes (which is the example with the metrics-processor case). Or maybe just partly which is why I'm kind of on the fence regarding the best tradeoff to pick.

Reply:
I see (I think I understand) what you mean. However what I hoped to achieve is that we will (almost) never need to update the `dbms_test.go` file to include new test cases. In the future, if we ever come across a SQL statement that we want to verify, we can simply create the test file pair (`.sql` and `.expected`) in `testdata` and the test suite will be able to pick it up and assert the output. 

The only way I can think of is to do something like below in the test output file
```go
// testdata/postgres/select/test-case-1.go
package "sqllexer"

var TestCaseExpected = output{
	Expected: "SELECT * FROM table;",
	// ... other fields ...
}

func init() {
  RegisterExpectedOutput("test-case", TestCaseExpected)
}
```
and in the `dbms_test.go` we do
```go
var ExpectedOutputs = map[string]output{}

func RegisterExpectedOutput(name string, out output) {
	ExpectedOutputs[name] = out
}
```

or maybe use `go generate`? not sure if that's a possible solution.

Reply:
> However what I hoped to achieve is that we will (almost) never need to update the dbms_test.go file to include new test cases. In the future, if we ever come across a SQL statement that we want to verify, we can simply create the test file pair (.sql and .expected) in testdata and the test suite will be able to pick it up and assert the output.

I think this works great if the test is doing simple input/output. Like if we had a sql input and a normalized/obfuscated output without anything else, it would work quite nicely. I think where the current version breaks down for me right now is that the expected files include the description of the expected output for that test case and it's more than just a plain input/output test. 

I think you're right that code generation isn't the way to go for this one. As for your alternative where you have the expected result defined in code, I guess you're trying to map the input sql files to their expected outputs but you could just have plain test cases as structs [like this](https://github.com/DataDog/dd-go/blob/prod/database-monitoring/apps/dbm-metadata-processor/decoder/decoder_test.go#L21-L27). You'd have your test cases defined as:
```go
testCases := []struct {		
		fixture         string // The name of the sql file relative to your embed filesystem
		expected    output
	} {
        // All test cases
        }
```

Yes, it requires touching the go test file to add new test cases but you've also got the IDE and compiler's help because it's all code. That doesn't seem bad to me but, you know, I just work here.

Reply:
This is what I get for defining the test cases in go. https://github.com/DataDog/go-sqllexer/compare/zhengda.lu/test-suite-2?expand=1
I had per DBMS test fixtures files so that new test case should only touch the dbms specific test file. 
Not sure if I like this, I kinda still like my old `json` approach better :smirk:

Reply:
I'm not sure if I like this one either but I had something different in mind. We should probably zoom and hack it. I suspect I'll see where things might not work well with what I have in mind or that we'll find something that works.